### ISSUE TITLE
fast mv between filesystem

### ISSUE THREAD

**Issue by derekdai** [openzfs/zfs#2991]
Is it possible to implement "fast mv" mechanism to move file/directory between filesystems in the same zpool without copy? I think it's kinda useful.

**Comment by behlendorf** [comment]
This is probably technically possible but I suspect it would be tricky to implement.  It would definitely be a nice feature to have, I've found myself wanting the same thing in the past.  Let's leave it open a feature request.

**Comment by derekdai** [comment]
Cool!

**Comment by GregorKopka** [comment]
While at it: if (maybe at some point in the future when this feature should be implemented) the code is able to relink a directory into another filesystem (mv), shouldn't it then not also be able to convert a directory into a new filesystem (and maybe the opposite, a filesystem into a directory of another, too)?

**Comment by DeHackEd** [comment]
Based on my understanding of how ZFS works internally, this would only be feasible under a few conditions:

* The file being moved must not be a directory
* The file must have only 1 reference (no other hard links)
* The file must not be present in the most recent snapshot of the filesystem.

It wouldn't really be possible under the general case.

**Comment by Feuermurmel** [comment]
Without understanding _any_ of the technical details involved here, I just had this idea and wanted to deposit it here:

Is it maybe possible to (mis)appropriate the deduplication feature for this? I.e. without having to enable deduplication for the pool in general, the same data structures could be created in the pool as would have been, if the original file and its copy were written with deduplication enabled? (I'm describing cross-fs-_copy_ here, think of mv as copy & delete, disregarding stuff like atomicity guarantees).

Making this more general, it would even allow what e.g. Btrfs calls _reflink-copies_ (moving or copying a slice of data between files without copying the data on disk). This would also be useful to quickly create a copy of a file within a single fs (this is the feature I'm actually looking for).

**Comment by GregorKopka** [comment]
@Feuermurmel This would work in case the data had been written with dedup in the first place, but only then - else the metadata of the original filesystem would miss the information of that data being deduplicated, a free from the original dataset would result in pool level data corruption as the additional datasets would point to then free space.

**Comment by fredcooke** [comment]
I agree that this would be nice to have, especially the way @GregorKopka described it. 

It seems very manual and awkward to do these kinds of things "the hard way" so to speak. Interesting reading some of the comments on how/why not etc, though.

**Comment by sporkman** [comment]
Looking for this in a situation where I have to move a gigantic Maildir-based mail spool. I cannot even begin to describe how painful moving a few TB of millions of files is on even a decent SSD-based pool. It would be such an amazing feature...

I know this request came in when this was the "zfsonlinux" repo, but perhaps some things have changed now that we're under the OpenZFS banner?

**Comment by Feuermurmel** [comment]
I assume the most likely candidate for a solution for this use case will be Block Reference Tables: [YouTube: Block Cloning by Pawel Jakub Dawidek](https://www.youtube.com/watch?v=qsE3R0Ysc8g)
